plugins/woocommerce/src/Blocks/Domain/Services/CheckoutFields.php

```diff
@@ -401,6 +401,55 @@ public function get_additional_fields() {
 		return $this->additional_fields;
 	}
 
+	/**
+	 * Gets the location of a field.
+	 *
+	 * @param string $field_key The key of the field to get the location for.
+	 * @return string The location of the field.
+	 */
+	public function get_field_location( $field_key ) {
+		foreach ( $this->fields_locations as $location => $fields ) {
+			if ( in_array( $field_key, $fields, true ) ) {
+				return $location;
+			}
+		}
+		return '';
+	}
+
+	/**
+	 * Validate an additional field against any custom validation rules. The result should be a WP_Error or true.
+	 *
+	 * @param true|\WP_Error $result       The current result of validating the field against the schema (before any custom validation has been applied).
+	 * @param string         $key          The key of the field.
+	 * @param mixed          $field_value  The value of the field.
+	 * @param array          $field_schema The schema of the field.
+	 *
+	 * @since 8.6.0
+	 */
+	public function validate_field( $result, $key, $field_value, $field_schema ) {
+
+		/**
+		 * Filter the result of validating an additional field.
+		 *
+		 * @param true|\WP_Error $result       The current result of validating the field.
+		 * @param mixed          $field_value  The value of the field.
+		 * @param array          $field_schema The schema of the field.
+		 * @param string         $key          The key of the field.
+		 *
+		 * @since 8.6.0
+		 */
+		$filtered_result = apply_filters( 'woocommerce_blocks_validate_additional_field_' . $key, $result, $field_value, $field_schema, $key );
```

Comment:
I was thinking more of an action instead of a filter, I understand you went with a filter because you want to check the return value.

However, does a filter work well here if you have more than 1 filter subscribed? what would happen if the first one returns an error, but the second one returns true? do we make it the consumer responsibility to first check `$result` type before deciding if they want to append to it (new error) or return it as is (no error), or return a new error.

An action misses the ability to return a value, but can very well throw a value, things like true, false, void, string, int, would be ignored, but an exception can be caught and appended, however, does an exception support passing over several hooks?

Reply:
Another way we can consider this to rename `$result` to $errors` so it's more explicit, it's a WP_Error with no errors on it, or maybe just an empty array that consumers can choose to:
- Do nothing (indicating validation passed).
- Append errors to it (indicating validation failed).

Reply:
Good point, I think it stops at the first error, which make sense, however, filters is still problematic here because it will run all validations and the last one might override the first one.

Reply:
I like the approach of passing an empty WP_Error and allowing extensions to add to it. I'll try with that.

Reply:
Consider the following problematic flow (which I dunno how likely it is thou)

If the GOV ID is above 5 letters, validation would pass fine.
```php
add_action(
	'woocommerce_loaded',
	function() {
		add_filter(
			'woocommerce_blocks_validate_additional_field_namespace/gov-id',
			function ( $result, $value, $schema, $key ) {
				if ( '12345' !== $value ) {
					return new WP_Error( 'invalid_gov_id', 'Sorry but only people with government ID: 12345 can order.' );
				}
				return $result;
			},
			10,
			4
		);

		add_filter(
			'woocommerce_blocks_validate_additional_field_namespace/gov-id',
			function ( $result, $value, $schema, $key ) {
				if ( strlen( $value ) > 5 ) {
					return true;
				} else {
					return new WP_Error( 'short_gov_id', 'Sorry but an ID should be more than 5 letters.' );
				}
			},
			11,
			4
		);
	}
);

```

Reply:
This is solved now by appending to an existing WP_Error.